godot_project/project_workspace/workspace_context/workspace_context.gd

```diff
@@ -1202,14 +1202,33 @@ func snapshot_moment(in_operation_name: String) -> void:
 	# workaround for part of the state for this workspace_context being inside ScriptUtils (this will apply that state)
 	get_editable_structure_contexts()
 	
-	# Apply simulation if the operation modified the structure of the project
-	if not History.is_operation_whitelisted(in_operation_name):
-		apply_simulation_if_running()
-	
 	# apply overdue signals, ensure snapshots contain up to date state
 	_emit_selection_in_structures_changed()
 	_emit_structure_content_changed()
 	
+	# Apply simulation if the operation modified the structure of the project
+	if not History.is_operation_whitelisted(in_operation_name) and is_simulating():
+		# We need to split the next snapshot in two separate steps:
+		# + Applying the simulation
+		# + Applying the user operation
+		# If we don't, hitting undo will revert the user action and the simulation at the same time.
+		
+		# Store the final snapshot containing both the applied state and the last action
+		var final_snapshot: Dictionary = _history.create_snapshot(in_operation_name)
+		var current_simulation_time: float = _simulation.get_last_seeked_time()
+		
+		# Go back just before starting the simulation
+		_history.apply_previous_snapshot()
+		
+		# Rewind simulation to the latest point and create a snapshot.
+		# This will drop the user operation (final_snapshot) from history. 
+		seek_simulation(current_simulation_time)
+		apply_simulation_if_running()
```

Comment:
I believe this logic is correct, but after executing this line I would expect for NanoStructure to have exactly the same state as it was before line 1217, can you remind me why this is not the case?

Reply:
this is not the case. in line 1217 is the sate of the simulation + the change of the user (in example create an atom)
in this case is the state of the simulation without the input of the user (without that extra atom)

In the past we used `NanoUndoRedo.create_action()` to detect an event was about to happen before it happened, this is no longer the case with the snapshot system, by the time you execute `snapshot_moment` the change is already made, so all of this is to rewind that last change.